activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb

```diff
@@ -11,14 +11,14 @@ def included(base) #:nodoc:
             :rollback_to_savepoint, :rollback_db_transaction, :exec_insert_all
 
           base.set_callback :checkout, :after, :configure_query_cache!
-          base.set_callback :checkin, :after, :disable_query_cache!
+          base.set_callback :checkin, :after, :clear_query_cache
         end
 
         def dirties_query_cache(base, *method_names)
           method_names.each do |method_name|
             base.class_eval <<-end_code, __FILE__, __LINE__ + 1
               def #{method_name}(*)
-                ActiveRecord::Base.clear_query_caches_for_current_thread if @query_cache_enabled
+                ActiveRecord::Base.clear_query_caches_for_current_thread if query_cache_enabled?
```

Comment:
This conditional was obviously already here, but does it sound right to you?

It seems surprising that wrapping an `uncached { }` block around a write operation can cause the cache to survive. :thinking:

Reply:
No, it doesn't.  IMO all writes should invalidate cache, but for some reason we have this conditional.  I'll do some research about why the conditional was added and post results here.

Reply:
Best I could find, it was introduced [here](https://github.com/rails/rails/commit/bfb906a905a#diff-c666cf7fe697634c32bc09b80a4d8787R19).  Maybe @jeremy remembers 12 year old commits? 🤣

I'll send another PR that removes the conditional and see what CI thinks about it.

Reply:
Sent a PR here: https://github.com/rails/rails/pull/37732